pkg/extract/redhat/csaf/csaf.go

```diff
@@ -680,216 +687,354 @@ func walkVulnerabilities(vulns []csaf.Vulnerability, pids []csaf.ProductID) (map
-						Operator: criteriaTypes.CriteriaOperatorTypeOR,
-						Criterions: []criterionTypes.Criterion{{
+
+	for cveID, apmm := range pmmm {
+		for ass, apm := range apmm {
+			for major, pmax := range apm {
+				ca := criteriaTypes.Criteria{
+					Operator: criteriaTypes.CriteriaOperatorTypeOR,
+				}
+
+				for _, p2 := range pmax.p2s {
+					vc, err := buildVersionCriterion(p2, ass)
+					if err != nil {
+						return nil, nil, errors.Wrap(err, "build version criterion")
+					}
+					if vc != nil {
+						ca.Criterions = append(ca.Criterions, criterionTypes.Criterion{
 							Type:    criterionTypes.CriterionTypeVersion,
-							Version: &vc,
-						}},
-					},
-					Tag: segmentTypes.DetectionTag(fmt.Sprintf("%s:%s", cve, pid)),
-				})
+							Version: vc,
+						})
+					}
+				}
+				conds := cm[major]
+				if len(ca.Criterions) > 0 {
+					conds = append(conds, conditionTypes.Condition{
+						Criteria: ca,
+						Tag:      calculateTag(cveID, pmax.maxPid),
+					})
+					cm[major] = conds
+				}
 			}
 		}
 	}
 
-	ds := make([]detectionTypes.Detection, 0, len(cm))
-	for v, cs := range cm {
+	var ds []detectionTypes.Detection
+	for major, conds := range cm {
 		ds = append(ds, detectionTypes.Detection{
-			Ecosystem:  ecosystemTypes.Ecosystem(fmt.Sprintf("%s:%s", ecosystemTypes.EcosystemTypeRedHat, v)),
-			Conditions: cs,
+			Ecosystem:  ecosystemTypes.Ecosystem(fmt.Sprintf("%s:%s", ecosystemTypes.EcosystemTypeRedHat, major)),
+			Conditions: conds,
 		})
 	}
 
-	as := []advisoryTypes.Advisory{{
-		Content: baseAdvisory,
-		Segments: func() []segmentTypes.Segment {
-			var ss []segmentTypes.Segment
-			for _, d := range ds {
-				for _, c := range d.Conditions {
-					ss = append(ss, segmentTypes.Segment{
-						Ecosystem: d.Ecosystem,
-						Tag:       c.Tag,
-					})
-				}
-			}
-			return ss
-		}(),
-	}}
+	return ds, pmmm, nil
+}
 
-	var vs []vulnerabilityTypes.Vulnerability
-	for _, vuln := range baseVulnerabilities {
-		for pid, ass := range vassm[string(vuln.ID)] {
-			ss, err := func() ([]severityTypes.Severity, error) {
-				var ss []severityTypes.Severity
-				if ass.severity.cvss2 != "" {
-					v2, err := cvssV2Types.Parse(ass.severity.cvss2)
-					if err != nil {
-						return nil, errors.Wrapf(err, "parse cvss2")
-					}
-					ss = append(ss, severityTypes.Severity{
-						Type:   severityTypes.SeverityTypeCVSSv2,
-						Source: "secalert@redhat.com",
-						CVSSv2: v2,
-					})
-				}
-				if ass.severity.cvss3 != "" {
-					switch {
-					case strings.HasPrefix(ass.severity.cvss3, "CVSS:3.0"):
-						v30, err := cvssV30Types.Parse(ass.severity.cvss3)
-						if err != nil {
-							return nil, errors.Wrap(err, "parse cvss3")
+func buildVersionCriterion(p2 product2, ass ass) (*vcTypes.Criterion, error) {
+	switch ass.status.product_status {
+	case "fixed":
+		if p2.name == "" && p2.version == "" {
+			return nil, nil
+		}
+
+		if len(p2.archs) == 1 && p2.archs[0] == "src" {
+			return nil, nil
+		}
```

Comment:
It doesn't have to be this PR, so this case seems like it should be an error when the product status is fixed.

Reply:
Issue created https://github.com/MaineK00n/vuls-data-update/issues/367